synapse/util/async_helpers.py

```diff
@@ -718,30 +719,25 @@ def failure_cb(val: Failure) -> None:
     return new_d
 
 
-# This class can't be generic because it uses slots with attrs.
-# See: https://github.com/python-attrs/attrs/issues/313
 @attr.s(slots=True, frozen=True, auto_attribs=True)
-class DoneAwaitable:  # should be: Generic[R]
+class DoneAwaitable(Awaitable[R]):
     """Simple awaitable that returns the provided value."""
 
-    value: Any  # should be: R
+    value: R
 
-    def __await__(self) -> Any:
-        return self
-
-    def __iter__(self) -> "DoneAwaitable":
-        return self
-
-    def __next__(self) -> None:
-        raise StopIteration(self.value)
+    def __await__(self) -> Generator[Any, None, R]:
+        yield None
```

Comment:
Not sure it matters, but is the yield less efficient?

I'm not 100% sure I understand removing `__iter__` and `__next__`.

Reply:
Err, being honest it's because I couldn't work out how to annotate the `__iter__` and `__next__` in a way that mypy was happy with. Can dig up the errors if you'd like.

I _think_ this is equivalent though:

- [`__await__` is supposed to return an iterator](https://docs.python.org/3/reference/datamodel.html?highlight=__await__#object.__await__)
- [an iterator is a thing with a `__next__`](https://docs.python.org/3/glossary.html#term-iterator)
-  (so was `__iter__` redundant`? Might have needed it before the `await` syntax added in Py 3.5?)

Reply:
Oh, I think it wasn't redundant because we were returning ourself, now `__await__` itself is an iterator?

Reply:
(Tbh we could probably use `asyncio.Future` instead of writing our own)

Reply:
> (Tbh we could probably use `asyncio.Future` instead of writing our own)

Alas, [`Future` docs](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future) say:

> Deprecated since version 3.10: Deprecation warning is emitted if loop is not specified and there is no running event loop.